src/v/coproc/pacemaker.cc

```diff
@@ -112,7 +114,24 @@ std::vector<errc> pacemaker::add_source(
       id, _shared_res, std::move(ctxs));
     const auto [itr, success] = _scripts.emplace(id, std::move(script_ctx));
     vassert(success, "Double coproc insert detected");
-    (void)itr->second->start();
+    vlog(coproclog.debug, "Adding source with id: {}", id);
+    (void)itr->second->start().handle_exception_type(
+      [this, id](const script_failed_exception& e) {
```

Comment:
I think there is a lifetime issue here. When the pacemaker is stopped it calls shutdown on each script context which waits on  the gate created in script_context::start() and started in the background here on the previous line. however, waiting on the gate doesn't wait for any continuation on the gate (i.e. this handle_exception_type). so during pacemaker shutdown i don't think that shutting down the script context background fibers acts as a barrier to prevent this exception handler from running after the pacemaker has shutdown. does that sound possible?

Reply:
Yes that is possible, however remove_source will be called which will attempt to lookup the script and not find it, then do nothing. Admittedly theres an issue if the lambda invokes this while the class is destructing.

Reply:
The point I'm trying to make is that the race looks like the pacemaker itself may have been destroyed... probably a shutdown scenario.

Reply:
you could probably put _all_ of it under the pacemaker's gate, or push the handler into the gate by passing it to start as a function object.

Reply:
Implemented in the latest force push, resolving, nice catch!